File: 1/2021.03.04/54.Spiral Matrix/java/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int m = matrix.length, n = matrix[0].length;
        int top = 0, bottom = m - 1, left = 0, right = n - 1;
        int[] dx = new int[]{0, 1, 0, -1};
        int[] dy = new int[]{1, 0, -1, 0};
        int x = 0, y = 0, di = 0;
        while (top <= bottom && left <= right) {
            res.add(matrix[x][y]);
            if (y + dy[di] > right) {
                top++;
                di = (di + 1) % 4;
            } else if (y + dy[di] < left) {
                bottom--;
                di = (di + 1) % 4;
            } else if (x + dx[di] > bottom) {
                right--;
                di = (di + 1) % 4;
            } else if (x + dx[di] < top) {
                left++;
                di = (di + 1) % 4;
            }
            x += dx[di];
            y += dy[di];
        }
    }
}

File: 1/2021.03.04/54.Spiral Matrix/python/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        dx = [0, 1, 0, -1]
        dy = [1, 0, -1, 0]
        x = y = di = 0
        res = []
        top, bottom, left, right = 0, m - 1, 0, n - 1
        while top <= bottom and left <= right:
            res.append(matrix[x][y])
            if y + dy[di] > right:
                di = (di + 1) % 4
                top += 1
            elif y + dy[di] < left:
                di = (di + 1) % 4
                bottom -= 1
            elif x + dx[di] > bottom:
                di = (di + 1) % 4
                right -= 1
            elif x + dx[di] < top:
                di = (di + 1) % 4
                left += 1
            x += dx[di]
            y += dy[di]
        return res
